docs/site/decorators/Decorators_inject.md

```diff
@@ -178,6 +178,39 @@ using `binding.to(value)`.
 To set other types of value providers such as `toDynamicValue`or `toClass`, use
 `@inject.binding` instead.
 
+### @instance
+
+`@instance` injects an instance of a given class from a binding created using
+`toClass()`.
+
+Syntax:
+`@instance(serviceClass?: Constructor, metadata: {skipSubClasses?: boolean})`.
+
+```ts
+class MyController {
+  constructor(@instance(MyService) public myService: MyService) {}
+}
+```
+
+Or even simpler:
+
+```ts
+class MyController {
+  // MyService is inferred from TypeScript design type
+  constructor(@instance() public myService: MyService) {}
+}
+```
+
+Please note the class has to be bound to the context with toClass().
+
+```ts
+const ctx = new Context();
+ctx.bind('my-service').toClass(MyService);
+```
+
+To prevent subclasses from being matched, set `skipSubClasses` to `false` for
+the `metadata` argument.
```

Comment:
As mentioned earlier, I find it potentially confusing that sub-classes are searched by default. How about starting with a stricter behavior where only the actual class is considered, and making sub-class search an opt-in feature? I.e. rename `skipSubClasses` to something like `searchSubClasses`.

@strongloop/loopback-next thoughts?

_I am cross-posting the original comment below._

---

I am concerned about the implications of this feature on user (developer) experience. Personally, I would find it surprising that `@service(MyService)` is resolved using the binding at key `services.MySubService`. If you look at the code wiring up the dependencies, there is no indication whatsoever that binding `services.MySubService` will automatically setup resolution of `services.MyService` too.

```ts
ctx.bind('services.MySubService').toClass(MySubService);
```

Let's find a better solution please. For example, we can make it explicit that one service class will be injected in places asking for another class.

```ts
// Resolve `MyService` service using `MySubService` implementation
ctx.service(MyService).toAlias(MySubService);

// Or maybe the other way around?
ctx.service(MySubService, {aliases: [MyService]})
```

Alternatively, let's keep handling of subclasses out of scope of this initial pull implementation and discuss it in a follow-up pull request.

Reply:
> We have a default Logger class from a component
> The application wants to have its own flavor.

First of all, I agree this is a valid use case that we should (eventually) support.

IMO, the application should bind `Logger` to `MyLogger` class in that case, to make it clear that we are overriding `Logger` to be resolved as `MyLogger`.

Both your examples are adding additional binding metadata that can be used by the consumer to look up the bound subclass, therefore they look slightly different from the problem I would like to discuss.

Let's take a look at the unit test from your pull request that I find problematic:

```ts
  it('injects an instance matching a sub class', async () => {
    class MyController {
      constructor(@instance(MyService) public myService: MyService) {}
    }
    ctx.unbind('services.MyService');
    ctx.bind('services.MySubService').toClass(MySubService);
    ctx.bind('controllers.MyController').toClass(MyController);
	
    const controller = await ctx.get<MyController>('controllers.MyController');
    expect(controller.myService).to.be.instanceOf(MySubService);
  });
```

When reading the code setting up bindings, it's not clear that `MyService` will be resolved to `MySubService` class.

I would like to have a more explicit setup, for example:

```ts
    ctx.bind('services.MyService').toClass(MySubService);
```

Another aspect I find potentially confusing - instance injection is comparing class constructors, it completely ignores binding keys. I believe the following setup would pass your test too:

```ts
    ctx.bind('repositories.MySubService').toClass(MySubService);
```

Is it a good idea to resolve `services.MyService` via `repositories.MySubService`?

The class-constructor check is also not very flexible. What if the application wants to use an async factory function to obtain the class instance? Because `@instance` is using class-based check, it won't be able to match a binding backed by a factory function (`.toDynamicValue()`).

I feel we need to revisit this pull request at high level. What scenario (user story) do we want to support & simplify by this new decorator? Are the extensibility options (e.g. subclass matching) worth the overhead in implementation complexity and the decrease in clarity of how are bindings resolved in a project using the extensibility options?